src/distributions/bernoulli.jl

```diff
@@ -34,14 +48,27 @@ function prod(::AddonProdLogScale, new_dist::Bernoulli, left_dist::Bernoulli, ri
     return log(a)
 end
 
-function prod(::AddonProdLogScale, new_dist::Bernoulli, left_dist::Bernoulli, right_dist::Categorical)
-    @assert length(probvec(right_dist)) === 2 "Improper Bernoulli x Categorical product"
-    left_p = succprob(left_dist)
-    right_p = first(probvec(right_dist))
-    a = left_p * right_p + (one(left_p) - left_p) * (one(right_p) - right_p)
-    return log(a)
+function prod(::AddonProdLogScale, new_dist::Categorical, left_dist::Bernoulli, right_dist::Categorical)
+
+    # get probability vectors
+    p_left = probvec(left_dist)
+    p_right = probvec(right_dist)
+
+    # find length of new vector and compute entries
+    if length(p_left) >= length(p_right)
```

Comment:
I think instead of if statement,  it  could be shorter to just write 
```
p_new = vcat(p_right..., zeros(abs(length(p_left) - length(p_right))))
Z = dot(p_left, p_new)

```

Reply:
nice suggestion, yet we either need to append zeros to `p_left` or `p_right` for which we can't get rid of the `if` statement

Reply:
@bartvanerp  I played around and this is what I came up with
```julia
function prod3(::ProdAnalytical, left::Bernoulli, right::Categorical)

    # get probability vectors
    p_left = probvec(left)
    p_right = probvec(right)

    # find the maximum length of both arguments
    mlength = max(length(p_left), length(p_right))
    
    # preallocate the result
    p_new = zeros(ReactiveMP.promote_samplefloattype(p_left, p_right), mlength)
    
    # Create 'e'xtended versions of the left and the right prob vectors
    # Do so by appending `0` with the Iterators.repeated, branch and allocation free
    e_left  = Iterators.flatten((p_left, Iterators.repeated(0, max(0, length(p_right) - length(p_left)))))
    e_right = Iterators.flatten((p_right, Iterators.repeated(0, max(0, length(p_left) - length(p_right)))))
    
    for (i, l, r) in zip(eachindex(p_new), e_left, e_right)
        @inbounds p_new[i] = l * r
    end

    # return categorical with normalized probability vector
    return Categorical(ReactiveMP.normalize!(p_new, 1))
end
```

```
@btime prod(ProdAnalytical(), $left, $right) # my version - 371.835 ns (5 allocations: 304 bytes)
@btime prod(ProdAnalytical(), $left, $right) # PR version - 487.253 ns (13 allocations: 496 bytes)
```

Reply:
Awesome!